File: backend/LiveSync.Api/Services/DocumentService.cs
using LiveSync.Api.Data;
using LiveSync.Api.DTOs;
using LiveSync.Api.Models;
using LiveSync.Execution.Contracts;
using Microsoft.EntityFrameworkCore;
using System.Security.Cryptography;

namespace LiveSync.Api.Services
{
    public class DocumentService : IDocumentService
    {
        private readonly ApplicationDbContext _context;
        private readonly ISandboxExecutionClient _sandboxExecutionClient;
        private readonly ILogger<DocumentService> _logger;

        public DocumentService(
            ApplicationDbContext context,
            ISandboxExecutionClient sandboxExecutionClient,
            ILogger<DocumentService> logger)
        {
            _context = context;
            _sandboxExecutionClient = sandboxExecutionClient;
            _logger = logger;
        }

        public async Task<DocumentDto?> GetDocumentByIdAsync(string documentId, string userId)
        {
            try
            {
                var document = await _context.Documents
                    .AsNoTracking()
                    .Include(d => d.Owner)
                    .Include(d => d.SharedWith)
                    .ThenInclude(s => s.User)
                    .FirstOrDefaultAsync(d => d.Id == documentId);

                if (document == null)
                    return null;

                // Check if user is owner or has shared access
                if (document.OwnerId != userId && !document.SharedWith.Any(s => s.UserId == userId))
                    return null;

                return MapToDto(document);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error getting document");
                return null;
            }
        }

        public Task<string?> GetAccessLevelAsync(string documentId, string userId)
        {
            return _context.Documents
                .AsNoTracking()
                .Where(d => d.Id == documentId)
                .Select(d => d.OwnerId == userId
                    ? "Edit"
                    : d.SharedWith
                        .Where(s => s.UserId == userId)
                        .Select(s => s.AccessLevel ?? "View")
                        .FirstOrDefault())
                .SingleOrDefaultAsync();
        }

        public async Task<List<DocumentDto>> GetUserDocumentsAsync(string userId)
        {
            try
            {
                var documents = await _context.Documents
                    .AsNoTracking()
                    .Include(d => d.Owner)
                    .Include(d => d.SharedWith)
                    .Where(d => d.OwnerId == userId)
                    .OrderByDescending(d => d.UpdatedAt)
                    .ToListAsync();

                return documents.Select(MapToDto).ToList();
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error getting user documents");
                return new List<DocumentDto>();
            }
        }

        public async Task<List<SharedDocumentDto>> GetSharedDocumentsAsync(string userId)
        {
            try
            {
                var sharedDocs = await _context.SharedDocuments
                    .AsNoTracking()
                    .Include(s => s.Document)
                    .ThenInclude(d => d!.Owner)
                    .Include(s => s.User)
                    .Where(s => s.UserId == userId)
                    .OrderByDescending(s => s.SharedAt)
                    .ToListAsync();

                return sharedDocs.Select(MapSharedToDto).ToList();
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error getting shared documents");
                return new List<SharedDocumentDto>();
            }
        }

        public async Task<DocumentDto> CreateDocumentAsync(string userId, CreateDocumentRequest request)
        {
            try
            {
                var document = new Document
                {
                    Title = request.Title,
                    Content = request.Content,
                    OwnerId = userId,
                    CreatedAt = DateTime.UtcNow,
                    UpdatedAt = DateTime.UtcNow
                };

                _context.Documents.Add(document);
                await _context.SaveChangesAsync();

                var user = await _context.Users.FindAsync(userId);
                if (user != null)
                {
                    document.Owner = user;
                }

                return MapToDto(document);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error creating document");
                throw;
            }
        }

        public async Task<DocumentDto?> UpdateDocumentAsync(string documentId, string userId, UpdateDocumentRequest request)
        {
            try
            {
                var document = await _context.Documents
                    .Include(d => d.Owner)
                    .Include(d => d.SharedWith)
                    .FirstOrDefaultAsync(d => d.Id == documentId);

                if (document == null)
                    return null;

                // Avoid a second database query after loading the access relationships.
                if (document.OwnerId != userId &&
                    !document.SharedWith.Any(s => s.UserId == userId && s.AccessLevel == "Edit"))
                    return null;

                if (request.Title is not null)
                    document.Title = request.Title.Trim();

                // Empty content is valid: clearing a document must be persisted.
                if (request.Content is not null)
                    document.Content = request.Content;

                document.UpdatedAt = DateTime.UtcNow;
                document.LastEditedAt = DateTime.UtcNow;
                document.LastEditedBy = request.LastEditedBy ?? userId;

                await _context.SaveChangesAsync();
                return MapToDto(document);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error updating document");
                return null;
            }
        }

        public async Task<bool> DeleteDocumentAsync(string documentId, string userId)
        {
            try
            {
                var document = await _context.Documents.FindAsync(documentId);

                if (document == null || document.OwnerId != userId)
                    return false;

                _context.Documents.Remove(document);
                await _context.SaveChangesAsync();
                return true;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error deleting document");
                return false;
            }
        }

        public async Task<DocumentDto?> UpdateContentAsync(string documentId, string userId, DocumentContentUpdateRequest request)
        {
            try
            {
                var document = await _context.Documents
                    .Include(d => d.Owner)
                    .Include(d => d.SharedWith)
                    .FirstOrDefaultAsync(d => d.Id == documentId);

                if (document == null)
                    return null;

                // Check if user has edit access
                if (!await HasEditAccessAsync(documentId, userId))
                    return null;

                document.Content = request.Content;
                document.UpdatedAt = DateTime.UtcNow;
                document.LastEditedAt = DateTime.UtcNow;
                document.LastEditedBy = request.LastEditedBy ?? userId;

                await _context.SaveChangesAsync();
                return MapToDto(document);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error updating content");
                return null;
            }
        }

        public async Task<IReadOnlyList<ExecutionLanguageDescriptor>> GetExecutionLanguagesAsync()
        {
            try
            {
                return await _sandboxExecutionClient.GetLanguagesAsync();
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error getting execution languages");
                return [];
            }
        }

        public async Task<DocumentExecutionResponse?> ExecuteDocumentAsync(string documentId, string userId, ExecuteDocumentRequest request)
        {
            try
            {
                var document = await _context.Documents
                    .AsNoTracking()
                    .FirstOrDefaultAsync(d => d.Id == documentId);

                if (document == null)
                    return null;

                if (!await HasEditAccessAsync(documentId, userId))
                    return null;

                var execution = await _sandboxExecutionClient.ExecuteAsync(new SandboxExecutionRequest
                {
                    Language = NormalizeExecutionLanguage(request.Language),
                    Code = document.Content,
                    StandardInput = request.StandardInput
                });

                return new DocumentExecutionResponse
                {
                    DocumentId = document.Id,
                    Language = execution.Language,
                    Status = execution.Status,
                    IsSuccess = execution.IsSuccess,
                    Message = execution.Message,
                    StandardOutput = execution.StandardOutput,
                    StandardError = execution.StandardError,
                    RequestedAt = execution.RequestedAt,
                    CompletedAt = execution.CompletedAt
                };
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error executing document");
                var timestamp = DateTime.UtcNow;
                return new DocumentExecutionResponse
                {
                    DocumentId = documentId,
                    Language = NormalizeExecutionLanguage(request.Language),
                    Status = "Failed",
                    IsSuccess = false,
                    Message = "Sandbox execution request failed.",
                    StandardError = ex.Message,
                    RequestedAt = timestamp,
                    CompletedAt = timestamp
                };
            }
        }

        public async Task<DocumentDto?> GenerateShareCodeAsync(string documentId, string userId)
        {
            try
            {
                var document = await _context.Documents
                    .Include(d => d.Owner)
                    .Include(d => d.SharedWith)
                    .FirstOrDefaultAsync(d => d.Id == documentId);

                if (document == null || document.OwnerId != userId)
                    return null;

                string shareCode;
                do
                {
                    shareCode = GenerateShareCode();
                }
                while (await _context.Documents.AnyAsync(d => d.ShareCode == shareCode));

                document.ShareCode = shareCode;
                await _context.SaveChangesAsync();
                return MapToDto(document);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error generating share code");
                return null;
            }
        }

        public async Task<DocumentDto?> GetDocumentByShareCodeAsync(string shareCode)
        {
            try
            {
                var document = await _context.Documents
                    .Include(d => d.Owner)
                    .Include(d => d.SharedWith)
                    .FirstOrDefaultAsync(d => d.ShareCode == shareCode);

                return document == null ? null : MapToDto(document);
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error getting document by share code");
                return null;
            }
        }

        public async Task<bool> AddSharedDocumentAsync(string shareCode, string userId)
        {
            try
            {
                // Find document by share code
                var document = await _context.Documents
                    .FirstOrDefaultAsync(d => d.ShareCode == shareCode);

                if (document == null)
                    return false;

                if (document.OwnerId == userId)
                    return false;

                // Check if user already has access
                var existingShare = await _context.SharedDocuments
                    .FirstOrDefaultAsync(s => s.DocumentId == document.Id && s.UserId == userId);

                if (existingShare != null)
                    return false; // Already has access

                // Use the document's default access level for the share code
                var accessLevel = document.DefaultAccessLevel ?? "View";

                // Create new shared document entry with the document's default access level
                var sharedDoc = new SharedDocument
                {
                    Id = Guid.NewGuid().ToString(),
                    DocumentId = document.Id,
                    UserId = userId,
                    SharedAt = DateTime.UtcNow,
                    AccessLevel = accessLevel
                };

                _context.SharedDocuments.Add(sharedDoc);
                await _context.SaveChangesAsync();

                return true;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error adding shared document");
                return false;
            }
        }

        public async Task<bool> RemoveSharedAccessAsync(string documentId, string userId, string sharedUserId)
        {
            try
            {
                var document = await _context.Documents.FindAsync(documentId);
                if (document == null || document.OwnerId != userId)
                    return false;

                var sharedDoc = await _context.SharedDocuments
                    .FirstOrDefaultAsync(s => s.DocumentId == documentId && s.UserId == sharedUserId);

                if (sharedDoc == null)
                    return false;

                _context.SharedDocuments.Remove(sharedDoc);
                await _context.SaveChangesAsync();
                return true;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error removing shared access");
                return false;
            }
        }

        public async Task<bool> HasEditAccessAsync(string documentId, string userId)
        {
            try
            {
                return await GetAccessLevelAsync(documentId, userId) == "Edit";
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error checking edit access");
                throw;
            }
        }

        public async Task<bool> UpdateSharedAccessLevelAsync(string documentId, string sharedUserId, string accessLevel)
        {
            try
            {
                // Validate access level
                if (accessLevel != "View" && accessLevel != "Edit")
                    return false;

                // Find the shared document entry
                var sharedDoc = await _context.SharedDocuments
                    .FirstOrDefaultAsync(s => s.DocumentId == documentId && s.UserId == sharedUserId);

                if (sharedDoc == null)
                    return false;

                // Update the access level
                sharedDoc.AccessLevel = accessLevel;
                await _context.SaveChangesAsync();

                return true;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error updating shared access level");
                return false;
            }
        }

        public async Task<bool> UpdateShareCodeAccessLevelAsync(string documentId, string userId, string accessLevel)
        {
            try
            {
                // Validate access level
                if (accessLevel != "View" && accessLevel != "Edit")
                    return false;

                // Find the document and verify ownership
                var document = await _context.Documents.FindAsync(documentId);
                if (document == null || document.OwnerId != userId)
                    return false;

                // Update the document's default access level
                document.DefaultAccessLevel = accessLevel;
                await _context.SaveChangesAsync();

                return true;
            }
            catch (Exception ex)
            {
                _logger.LogError(ex, "Error updating share code access level");
                return false;
            }
        }

        public string GenerateShareCode()
        {
            const string chars = "ABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789";
            return RandomNumberGenerator.GetString(chars, 10);
        }

        private static string NormalizeExecutionLanguage(string language)
        {
            var normalized = language.Trim().ToLowerInvariant();
            return normalized == "cs" ? "csharp" : normalized;
        }

        private DocumentDto MapToDto(Document document)
        {
            return new DocumentDto
            {
                Id = document.Id,
                Title = document.Title,
                Content = document.Content,
                OwnerId = document.OwnerId,
                OwnerName = document.Owner?.UserName ?? "Unknown",
                ShareCode = document.ShareCode,
                DefaultAccessLevel = document.DefaultAccessLevel,
                CreatedAt = document.CreatedAt,
                UpdatedAt = document.UpdatedAt,
                LastEditedAt = document.LastEditedAt,
                LastEditedBy = document.LastEditedBy,
                SharedWith = document.SharedWith?.Select(s => new SharedDocumentDto
                {
                    Id = s.Id,
                    DocumentId = s.DocumentId,
                    UserId = s.UserId,
                    UserName = s.User?.UserName ?? "Unknown",
                    SharedAt = s.SharedAt,
                    AccessLevel = s.AccessLevel ?? "View"
                }).ToList() ?? new List<SharedDocumentDto>()
            };
        }

        private SharedDocumentDto MapSharedToDto(SharedDocument shared)
        {
            return new SharedDocumentDto
            {
                Id = shared.Id,
                DocumentId = shared.DocumentId,
                DocumentTitle = shared.Document?.Title ?? "Unknown",
                UserId = shared.UserId,
                UserName = shared.User?.UserName ?? "Unknown",
                SharedAt = shared.SharedAt,
                AccessLevel = shared.AccessLevel ?? "View"
            };
        }
    }
}

File: frontend/src/app/services/document.service.ts
import { Injectable, inject } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { firstValueFrom } from 'rxjs';
import { appEndpoints } from '../app-endpoints';

export interface DocumentDto {
  id: string;
  title: string;
  content: string;
  ownerId: string;
  ownerName?: string;
  shareCode?: string;
  defaultAccessLevel: string;
  createdAt: string;
  updatedAt: string;
  lastEditedAt?: string;
  lastEditedBy?: string;
  sharedWith: SharedDocumentDto[];
}

export interface SharedDocumentDto {
  id: string;
  documentId: string;
  documentTitle: string;
  userId: string;
  userName?: string;
  sharedAt: string;
  accessLevel: string;
}

export interface CreateDocumentRequest {
  title: string;
  content?: string;
}

export interface UpdateDocumentRequest {
  title?: string;
  content?: string;
  lastEditedBy?: string;
}

export interface DocumentContentUpdateRequest {
  content: string;
  lastEditedBy?: string;
}

export interface DocumentAccessResponse {
  accessLevel: string;
}

export interface ExecuteDocumentRequest {
  language: string;
  standardInput?: string;
}

export interface DocumentExecutionResponse {
  documentId: string;
  language: string;
  status: string;
  isSuccess: boolean;
  message: string;
  standardOutput?: string;
  standardError?: string;
  requestedAt: string;
  completedAt: string;
}

interface MessageResponse {
  message: string;
}

@Injectable({
  providedIn: 'root',
})
export class DocumentService {
  private readonly http = inject(HttpClient);
  private readonly apiUrl = `${appEndpoints.apiBaseUrl}/api/documents`;

  async getMyDocuments(): Promise<DocumentDto[]> {
    try {
      const response = await firstValueFrom(
        this.http.get<DocumentDto[]>(`${this.apiUrl}/my-documents`),
      );
      return response;
    } catch (error) {
      console.error('Error fetching documents:', error);
      throw error;
    }
  }

  async getSharedDocuments(): Promise<SharedDocumentDto[]> {
    try {
      const response = await firstValueFrom(
        this.http.get<SharedDocumentDto[]>(`${this.apiUrl}/shared-with-me`),
      );
      return response;
    } catch (error) {
      console.error('Error fetching shared documents:', error);
      throw error;
    }
  }

  async getDocument(id: string): Promise<DocumentDto> {
    try {
      const response = await firstValueFrom(this.http.get<DocumentDto>(`${this.apiUrl}/${id}`));
      return response;
    } catch (error) {
      console.error('Error fetching document:', error);
      throw error;
    }
  }

  async getAccessLevel(id: string): Promise<string> {
    try {
      const response = await firstValueFrom(
        this.http.get<DocumentAccessResponse>(`${this.apiUrl}/${id}/access`),
      );
      return response.accessLevel;
    } catch (error) {
      console.error('Error fetching document access:', error);
      throw error;
    }
  }

  async createDocument(request: CreateDocumentRequest): Promise<DocumentDto> {
    try {
      const response = await firstValueFrom(this.http.post<DocumentDto>(`${this.apiUrl}`, request));
      return response;
    } catch (error) {
      console.error('Error creating document:', error);
      throw error;
    }
  }

  async updateDocument(id: string, request: UpdateDocumentRequest): Promise<DocumentDto> {
    try {
      const response = await firstValueFrom(
        this.http.put<DocumentDto>(`${this.apiUrl}/${id}`, request),
      );
      return response;
    } catch (error) {
      console.error('Error updating document:', error);
      throw error;
    }
  }

  async updateContent(id: string, request: DocumentContentUpdateRequest): Promise<DocumentDto> {
    try {
      const response = await firstValueFrom(
        this.http.put<DocumentDto>(`${this.apiUrl}/${id}/content`, request),
      );
      return response;
    } catch (error: any) {
      console.error('Error updating content:', error);
      // Add specific handling for permission errors
      if (error.status === 401 || error.status === 403) {
        const permissionError = new Error(
          'Permission denied: You no longer have edit access to this document',
        );
        (permissionError as any).status = error.status;
        (permissionError as any).isPermissionError = true;
        throw permissionError;
      }
      throw error;
    }
  }

  async executeDocument(
    id: string,
    request: ExecuteDocumentRequest,
  ): Promise<DocumentExecutionResponse> {
    try {
      return await firstValueFrom(
        this.http.post<DocumentExecutionResponse>(`${this.apiUrl}/${id}/execute`, request),
      );
    } catch (error) {
      console.error('Error executing document:', error);
      throw error;
    }
  }

  async getExecutionLanguages(): Promise<string[]> {
    try {
      return await firstValueFrom(this.http.get<string[]>(`${this.apiUrl}/execution-languages`));
    } catch (error) {
      console.error('Error fetching execution languages:', error);
      throw error;
    }
  }

  async deleteDocument(id: string): Promise<void> {
    try {
      await firstValueFrom(this.http.delete(`${this.apiUrl}/${id}`));
    } catch (error) {
      console.error('Error deleting document:', error);
      throw error;
    }
  }

  async generateShareCode(id: string): Promise<DocumentDto> {
    try {
      const response = await firstValueFrom(
        this.http.post<DocumentDto>(`${this.apiUrl}/${id}/generate-share-code`, {}),
      );
      return response;
    } catch (error) {
      console.error('Error generating share code:', error);
      throw error;
    }
  }

  async getDocumentByShareCode(shareCode: string): Promise<DocumentDto> {
    try {
      const response = await firstValueFrom(
        this.http.get<DocumentDto>(`${this.apiUrl}/share/${shareCode}`),
      );
      return response;
    } catch (error) {
      console.error('Error fetching document by share code:', error);
      throw error;
    }
  }

  async addSharedDocument(shareCode: string): Promise<MessageResponse> {
    try {
      return await firstValueFrom(
        this.http.post<MessageResponse>(`${this.apiUrl}/add-shared`, { shareCode }),
      );
    } catch (error) {
      console.error('Error adding shared document:', error);
      throw error;
    }
  }

  async removeSharedAccess(documentId: string, sharedUserId: string): Promise<void> {
    try {
      await firstValueFrom(this.http.delete(`${this.apiUrl}/${documentId}/shared/${sharedUserId}`));
    } catch (error) {
      console.error('Error removing shared access:', error);
      throw error;
    }
  }

  async updateSharedAccessLevel(
    documentId: string,
    sharedUserId: string,
    accessLevel: string,
  ): Promise<MessageResponse> {
    try {
      return await firstValueFrom(
        this.http.put<MessageResponse>(
          `${this.apiUrl}/${documentId}/shared/${sharedUserId}/access-level`,
          { accessLevel },
        ),
      );
    } catch (error) {
      console.error('Error updating access level:', error);
      throw error;
    }
  }

  async updateShareCodeAccessLevel(
    documentId: string,
    accessLevel: string,
  ): Promise<MessageResponse> {
    try {
      return await firstValueFrom(
        this.http.put<MessageResponse>(`${this.apiUrl}/${documentId}/share-code-access-level`, {
          accessLevel,
        }),
      );
    } catch (error) {
      console.error('Error updating share code access level:', error);
      throw error;
    }
  }
}
